rclcpp/include/rclcpp/buffers/buffer_implementation_base.hpp

```diff
@@ -0,0 +1,37 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__BUFFERS__BUFFER_IMPLEMENTATION_BASE_HPP_
+#define RCLCPP__BUFFERS__BUFFER_IMPLEMENTATION_BASE_HPP_
+
+namespace rclcpp
+{
+namespace intra_process_buffer
+{
+
+template<typename BufferT>
+class BufferImplementationBase
+{
+public:
+  virtual void dequeue(BufferT & request) = 0;
```

Comment:
Why just not returning `BufferT`? RVO performs better than using `std::move`

Reply:
I created a branch where I allow RVO https://github.com/alsora/rclcpp/commit/68ef5cba54442894a4b201a1ee3045aaa8f30553
However, the performances are slightly worst

Reply:
I created a branch where I allow RVO. 
Here are the differences https://github.com/alsora/rclcpp/compare/alsora/simplified_ipc...alsora:alsora/rvalue_optimization

The performances are slightly better than the current version (in 10 runs of MontBlanc benchmark application we have 178us latency and 11.3% CPU in the current against 168us latency and 10.5% CPU in the other).

Is it the suggested behavior to rely on the compiler optimization in this case?

Reply:
RVO or "copy elision" became part of the C+11 standard, so it's safe to rely on it.

Reply:
Correction: _copy elision_ is not guaranteed by pre C++17 ([relevant](https://blog.tartanllama.xyz/guaranteed-copy-elision/)) standards but only allowed. However, most modern compilers implement it.

Reply:
Merged